rfcs/62-editor-data-model.md

```diff
@@ -0,0 +1,555 @@
+# Feature Name: `editor_data_model`
+
+## Summary
+
+This RFC describes the data model and API that the Editor uses in memory, and the basis for communicating with the Game process and for serializing that data to disk (although the serialization format to disk is outside of the scope of this RFC). The data model is largely inspired by the one of [Our Machinery](https://ourmachinery.com/), as described in [this blog post](https://ourmachinery.com/post/the-story-behind-the-truth-designing-a-data-model/).
+
+## Glossary
+
+In this RFC, the following terms are used:
+
+- **Game** refers to any application made with Bevy, which encapsulates proper games but also any other kind of Bevy application (CAD, architecture software, _etc._).
+- **Author** refers to any person (including non-technical) building the Game with the intent of shipping it to end users.
+- **Editor** refers to the software used to edit Game data and prepare them for runtime consumption, and which this RFC describes the data model of.
+- **Runtime** refers to the Game or Editor process running time, as opposed to compiling or building time. When not otherwise specified, this refers to the Game's runtime. In general it's synonymouns of a process running, and is opposed to _build time_ or _compile time_.
+
+## Motivation
+
+We need a data model for the Bevy Editor. We want to make it decoupled from the runtime data model that the Game uses, which is based on compiled Rust types (`World`, _etc._) with a defined native layout from a fixed Rust compiler version (as most types don't use `#[repr(C)]`), so we can handle data migration across Rust compiler versions and therefore across both Editor and Game versions. We also want to make it _transportable_, that is serializable with the intent to transport it over any IPC mechanism, so that the Editor process can dialog with the Game process to enable real-time Game editing ("Play Mode").

```

Comment:
> What does native layout mean? We should probably use a clearer expression.

Machine layout? Binary layout?

> Readers just need to know that we're required to transport data from the Editor back to the Game through IPC.

Readers here are RFC reviewers right? If so, I'd argue they need to understand why we use IPC.

Reply:
>> What does native layout mean? We should probably use a clearer expression.
>
> Machine layout? Binary layout?

The problem is that it's difficult to understand what you're referring about here. The thing is that all data structures are just an abstraction introduced by high level programming language, and this abstraction is lost just right when the code is translated into assembly, where the fields of the data structure defined in the high level programming language are accessed with a base pointer + offset method. (Probably not 100% correct since i don't have much experience here, but the gist is that)

Personally, I would just avoid talking about machine code representations of code, when possible. The reason that we want the Editor to have a different data model from the Runtime's is because they, by our requirement, do not depend on each other. The machine-code incompatibility between them is just a consequence of that.

>> Readers just need to know that we're required to transport data from the Editor back to the Game through IPC.
>
> Readers here are RFC reviewers right? If so, I'd argue they need to understand why we use IPC.

Who reviews the RFC is just required to know the current state of the project, the already merged RFCs and what has been defined in the RFC up to that point, so they may not know about concepts developed in the bevyengine/bevy#5043 discussion.

However, in that case, I didn't ask to add information, but to remove some, because it seems not relevant to the RFC. We need a data model that supports IPC between the Editor and the Game (because we want them to be independent from each other), but it seems inappropriate to talk about the “Play Mode” here, where it's not mentioned anywhere else apart from the Unity prior art section.

Reply:
> The thing is that all data structures are just an abstraction introduced by high level programming language, and this abstraction is lost just right when the code is translated into assembly, where the fields of the data structure defined in the high level programming language are accessed with a base pointer + offset method.

The native layout is the offsets. This is the part which is not lost. I don't have a better term for it, that's the one I commonly use and hear used to talk about this.

> Personally, I would just avoid talking about machine code representations of code, when possible. The reason that we want the Editor to have a different data model from the Runtime's is because they, by our requirement, do not depend on each other. The machine-code incompatibility between them is just a consequence of that.

Fair enough. The native layout will probably be relevant for another RFC but not this one, will remove.

> but it seems inappropriate to talk about the “Play Mode” here

OK sorry I thought the issue was with the mention of IPC, not about Play Mode. I can remove that one.

Reply:
Removed both native layout and Play Mode references.